### ISSUE TITLE
Ability to derive spans for functions

### ISSUE THREAD

**Issue by hawkw** [hawkw/tokio-trace-prototype#9]
It seems like it should be possible (and perhaps not terribly difficult) to annotate functions so that a span is created and automatically entered for every call to that function, with the arguments added as fields on that span.

For example, something like
```rust
#[derive(Span)]
fn my_function(foo: usize, bar: String) -> String {
    // function body ...
}
```
could expand to
```rust
fn my_function(foo: usize, bar: String) -> String {
    span!("my_function", foo = foo, bar = bar.clone())
        .enter(move || {
            // original function body pre-expansion
        })
}
```

I think this expansion would be fairly easy to implement, and could even be accomplished with just a `macro_rules!` rather than a procedural macro (though then we wouldn't get the nice `#[derive(...)]` attribute). 

One main caveat: this would only work for functions whose arguments are valid field `Value`s, and implement `Clone`, since we can't pass them by value into the span's fields and still allow the function body to use them, otherwise. The macro (even if it's a procedural macro) can't easily detect this, so we'd have to output incorrect code and let the compiler complain about it. 

A potential workaround to this would be to introduce another attribute that can be added to function arguments to tell the macro that they should not be added as fields on the span, allowing functions which have one or more `!Clone` argument to still derive spans. Alternatively, we could allow the derive attribute to accept a whitelist of fields which should be added to the trace (which might potentially be easier to implement).

**Comment by hawkw** [comment]
As a bit of a stretch goal, we could potentially also generate events or new spans on reassignments to values, like https://github.com/dtolnay/syn/tree/master/examples/trace-var. I'm not sure whether something that granular is worth the effort or not, though --- and it would take some thought regarding how reassignments fit into the current immutable span model...

**Comment by davidbarsky** [comment]
Couple assorted thoughts/capturing assorted thoughts:

- The clone requirements exists because a span could outlive a function parameter. This could _maybe_ worked around by a requiring a `Display` instead of a clone, thereby storing a string representation of the value.
- Given that 1.30 is landing in 4 weeks, an attribute macro might be sufficient for functions. an trace/instrument derive could also exist on structs that eventually implement Tower’s service trait.

**Comment by hawkw** [comment]
@davidbarsky:
> The clone requirements exists because a span could outlive a function parameter. This could _maybe_ worked around by a requiring a `Display` instead of a clone, thereby storing a string representation of the value. However, this does lose _some_ data.

Forgive me if I'm misunderstanding what you're saying here. Are you saying that the `Clone` requirement could be avoided by eagerly formatting the fields when the span is constructed? Because that's *true*, but I don't think it's a good idea. 

Up until this point, I've taken it as a design constraint that field values are lazily formatted. Here is some of the motivation behind that:
1. It should be the `Subscriber`'s choice as to how spans are formatted. Eventually, subscribers should be able to try to downcast values as implementors of `Debug` or `Display` or `serde::Value` and choose what implementation to use for seriallizing them.
2. It's possible that all the subscribers that see a given span may not care about the value of a particular field, or about the value of _any_ fields --- they might only be tracking data like span name, source code location, and enter/exit time, for example. In that case, it seems like it's better to avoid doing the work of serializing all the span's fields when they may never actually be used.
3. Something else I've taken as a design goal is that *constructing* a span or event should be cheap, since it might occur in a hot path. For example, in a production implementation, you might want to write a `Subscriber` that forwards all the spans and events it sees to a consumer in a dedicated thread for processing, getting the event processing work out of the performance-critical threads as soon as possible. Therefore, it's best to avoid doing work at span construction time, if possible.

Those are my high level thoughts on why I think lazily formatting span values is important. I'm open to discuss any of the design constraints if I'm misunderstanding the problem space!

Additionally, I think we *probably* want to avoid replacing the `Clone` requirement with `Display` on derived spans for function arguments for the following reasons:
1. Most function arguments are probably not `Display`. I think `Clone` is probably much more likely to occur on function arguments.
2. As I mentioned above, I think span construction should be cheap. Admittedly, cloning something like a `String` is not necessarily cheap. However, if it's well-documented that deriving a span for the arguments of a function is opting in to cloning them, users might recognize and `Arc` arguments that are otherwise slow to clone. A ref-count bump is almost certainly much faster than a majority of `Display` implementations...
3. Overall, I don't know if I agree that `Display` is the right formatter to use for trace data in general...

WDYT?

**Comment by davidbarsky** [comment]
While working on the span `#[trace]` macro, I was stumped—why is a `span("example call")` inside a `quote` breaking with this error?

```
    Checking example v0.1.0 (/Users/dbarsky/Developer/Rust/tokio-trace-prototype/tokio-trace-macros/example)
error: unexpected end of macro invocation
  --> tokio-trace-macros/example/src/main.rs:19:22
   |
19 |     let span = span!("example_span");
   |                      ^^^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `example`.
```

Turns out, `span!` requires _at least_ one key-value pair to be passed after the span name. Maybe that restriction should be raised...

**Comment by hawkw** [comment]
@davidbarsky 
> Turns out, span! requires at least one key-value pair to be passed after the span name. Maybe that restriction should be raised...

A trailing comma will also work, IIRC.

This is a bit of a rough spot in how the macro is parsed...I need to fix it but forgot to open an issue.

**Comment by davidbarsky** [comment]
No worries! I wonder if you're gonna expand effort to fix that bug, I wonder if it'll make sense to re-write it as a [function-like proc macro](https://doc.rust-lang.org/nightly/reference/procedural-macros.html#function-like-procedural-macros).

(Mainly—proc macros have _far_ nicer parsing errors, but that should be weighed against effort learning them/other priorities.)

**Comment by hawkw** [comment]
I don't think it requires much effort --- fairly sure it's a 1-character change, actually.

**Comment by hawkw** [comment]
@davidbarsky ef2edd1b95d08260aaf14e8050184abde07b1509 should fix the issue with the `span!` macro. :)

**Comment by davidbarsky** [comment]
Nice! I didn’t realize how simple the solution would be!

**Comment by hawkw** [comment]
yeah --- I was *hoping* I could fix it by moving the comma one character to the left, but I had to go with the brute Force solution. 

possibly when this crate becomes more production-ready, it will be worth replacing this with a proc macro for better parse errors, but I think for prototyping purposes the macro by example is significantly simpler in this case.

**Comment by hawkw** [comment]
Closed by #16.
